### PR TITLE
Add OpenAPI/Swagger support for minimal actions

### PR DESCRIPTION
**Pull request by halter73**
image

Addresses #30662

### CODE REVIEW

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+        public void OnProvidersExecuting(ApiDescriptionProviderContext context)
+        {
+            foreach (var endpoint in _endpointDataSource.Endpoints)
+            {
+                if (endpoint is RouteEndpoint routeEndpoint
+                    && routeEndpoint.Metadata.GetMetadata<MethodInfo>() is { } methodInfo
+                    && routeEndpoint.Metadata.GetMetadata<IHttpMethodMetadata>() is { } httpMethodMetadata)
+                {
+                    // REVIEW: Should we add an ApiDescription for endpoints without IHttpMethodMetadata? Swagger doesn't handle
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs:40]: Interesting q - MVC does add an entry for items that do not specify a method constraint (with a null value), so maybe it's fine to add one: https://github.com/dotnet/aspnetcore/blob/44bcd96cc93131bdcf184bd7e22364c7dab18a96/src/Mvc/Mvc.ApiExplorer/src/DefaultApiDescriptionProvider.cs#L341-L351

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            {
+                return (BindingSource.Body, parameter.Name ?? string.Empty);
+            }
+        }
+
+        private static ApiResponseType? CreateApiResponseType(Type responseType)
+        {
+            if (typeof(IResult).IsAssignableFrom(responseType))
+            {
+                // Can't determine anything about IResults yet. IResult<T> could help here.
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs:178]: We could treat this case the same as void returning?

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+        private static ApiParameterDescription CreateApiParameterDescription(ParameterInfo parameter, RoutePattern pattern)
+        {
+            var parameterType = parameter.ParameterType;
+
+            var (source, name) = GetBindingSourceAndName(parameter, pattern);
+
+            return new ApiParameterDescription
+            {
+                Name = name,
+                ModelMetadata = new EndpointMethodInfoModelMetadata(ModelMetadataIdentity.ForType(parameterType)),
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs:121]: Seems unfortunate we have to implement this. It doesn't look like there's anything other than the mdoel type that Swashbuckle reads from ModelMetadata.

**Comment by bradygaster** [comment]
OH SO EXCITING

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs (hunk 1 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs:40]: I tried null initially since `ApiDescription.HttpMethod` is indeed nullable, but I get this when trying this with Swashbuckle 5.6.3.

```
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Swashbuckle.AspNetCore.SwaggerGen.SwaggerGeneratorException: Ambiguous HTTP method for action - . Actions require an explicit HttpMethod binding for Swagger/OpenAPI 3.0
         at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GenerateOperations(IEnumerable`1 apiDescriptions, SchemaRepository schemaRepository) in Swashbuckle.AspNetCore.SwaggerGen.dll:token 0x60000f0+0x76
         at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GeneratePaths(IEnumerable`1 apiDescriptions, SchemaRepository schemaRepository) in Swashbuckle.AspNetCore.SwaggerGen.dll:token 0x60000ef+0x60
         at Swashbuckle.AspNetCore.SwaggerGen.SwaggerGenerator.GetSwagger(String documentName, String host, String basePath) in Swashbuckle.AspNetCore.SwaggerGen.dll:token 0x60000ed+0xe0
         at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider) in Swashbuckle.AspNetCore.Swagger.dll:token 0x6000009+0xa9
         at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application) in Microsoft.AspNetCore.Server.Kestrel.Core.dll:token 0x6000a9f+0x1b8
```

Is there a workaround?

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        {
+            var apiDescription = new ApiDescription
+            {
+                HttpMethod = httpMethod,
+                RelativePath = pattern.RawText?.TrimStart('/'),
+                ActionDescriptor = new ActionDescriptor
+                {
+                    RouteValues =
+                    {
+                        // Swagger uses this to group endpoints together.
```
Comment (davidfowl) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMethodInfoApiDescriptionProvider.cs:65]: Swashbucklet uses this? What a hack 😄

**Review by bradygaster — APPROVED** [review]

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+                return (BindingSource.Header, headerAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromBodyMetadata).IsAssignableFrom(a.AttributeType)))
+            {
+                return (BindingSource.Body, parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromServiceMetadata).IsAssignableFrom(a.AttributeType)) ||
+                     parameter.ParameterType == typeof(HttpContext) ||
+                     parameter.ParameterType == typeof(CancellationToken) ||
+                     parameter.ParameterType.IsInterface)
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:146]: TODO: rebase and copy the new `IServiceProviderIsService` logic used in RequestDelegateFactory. Unfortunately this breaks my out-of-repo sample, since we don't have a full SDK with this changes available from dotnet/installers yet 😢

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+                    MediaType = contentType,
+                });
+            }
+        }
+
+        // The CompilerGeneratedAttribute doesn't always get added so we also check if the type name starts with "<"
+        // For example,w "<>c" is a "declaring" type the C# compiler will generate without the attribute for a top-level lambda
+        // REVIEW: Is there a better way to do this?
+        private static bool IsCompilerGenerated(Type type) =>
+            Attribute.IsDefined(type, typeof(CompilerGeneratedAttribute)) || type.Name.StartsWith('<');
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:311]: @jaredpar Is there a better way to do this?

File: src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs [hunk 7]
```diff
@@ -77,13 +77,48 @@ private static List<IApiResponseMetadataProvider> GetResponseMetadataAttributes(
...
+
+            return responseTypes;
+        }
+
+        // Shared with EndpointMetadataApiDescriptionProvider
+        internal static List<ApiResponseType> ReadResponseMetadata(
+            IReadOnlyList<IApiResponseMetadataProvider> responseMetadataAttributes,
+            Type? type,
+            Type defaultErrorType,
+            MediaTypeCollection contentTypes)
```
Comment (halter73) [inline src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs:115]: @pranavkm I decided to share logic for determining response formats with `ApiResponseTypeProvider`. I don't think I changed the behavior.

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+            if (methodInfo.DeclaringType is not null && !IsCompilerGenerated(methodInfo.DeclaringType))
+            {
+                controllerName = methodInfo.DeclaringType.Name;
+            }
+            else
+            {
+                // If the declaring type is null or compiler-generated (e.g. lambdas),
+                // group the methods under a "Map" controller.
+                controllerName = "Map";
```
Comment (davidfowl) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:79]: A better fallback would be the application name. `IHostEnvironment.ApplicationName` or the assembly entry point name.

File: src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    /// <summary>
+    /// 
+    /// </summary>
+    public static class EndpointMetadataApiExplorerServiceCollectionExtensions
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="services"></param>
+        public static void AddEndpointMetadataApiExplorer(this IServiceCollection services)
```
Comment (davidfowl) [inline src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs:19]: This name feels a little literal.

File: src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs (hunk 7 above)
Comment (pranavkm) [reply src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs:115]: No issues. This type has enough unit tests so as long as they continue passing, this should be fine.

File: src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    /// <summary>
+    /// Extensions for configuring ApiExplorer using <see cref="Endpoint.Metadata"/>.
+    /// </summary>
+    public static class EndpointMetadataApiExplorerServiceCollectionExtensions
+    {
+        /// <summary>
+        /// Configures ApiExplorer using <see cref="Endpoint.Metadata"/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        public static void AddEndpointsApiExplorer(this IServiceCollection services)
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs:20]: Could we do a super quick email review for the name? See if others have a more compelling name suggestion? This is alright, but doesn't feel great.

Comment (pranavkm) [reply src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs:20]:
```suggestion
        public static IServiceCollection AddEndpointsApiExplorer(this IServiceCollection services)
```

It matches other Add* APIs

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    internal class EndpointMetadataApiDescriptionProvider : IApiDescriptionProvider
+    {
+        private readonly EndpointDataSource _endpointDataSource;
+        private readonly IHostEnvironment _environment;
+        private readonly IServiceProviderIsService? _serviceProviderIsService;
+
+        // Executes before MVC's DefaultApiDescriptionProvider and GrpcHttpApiDescriptionProvider for no particular reason :D
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:30]:
```suggestion
        // Executes before MVC's DefaultApiDescriptionProvider and GrpcHttpApiDescriptionProvider for no particular reason
```

UTF-8 or bust

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+            if (methodInfo.DeclaringType is not null && !IsCompilerGenerated(methodInfo.DeclaringType))
+            {
+                controllerName = methodInfo.DeclaringType.Name;
+            }
+            else
+            {
+                // If the declaring type is null or compiler-generated (e.g. lambdas),
+                // group the methods under a "Map" controller.
+                controllerName = _environment.ApplicationName;
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:85]: Out of curiosity, does leaving it empty work? Would we want that?

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+            {
+                Name = name,
+                ModelMetadata = CreateModelMetadata(parameter.ParameterType),
+                Source = source,
+                DefaultValue = parameter.DefaultValue,
+                Type = parameter.ParameterType,
+            };
+        }
+
+        // TODO: Share more of this logic with RequestDelegateFactory.CreateArgument(...) using RequestDelegateFactoryUtilities
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:135]: Could you file an issue and remove this?

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+            }
+            else if (attributes.OfType<IFromQueryMetadata>().FirstOrDefault() is { } queryAttribute)
+            {
+                return (BindingSource.Query, queryAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (attributes.OfType<IFromHeaderMetadata>().FirstOrDefault() is { } headerAttribute)
+            {
+                return (BindingSource.Header, headerAttribute.Name ?? parameter.Name ?? string.Empty);
+            }
+            else if (parameter.CustomAttributes.Any(a => typeof(IFromBodyMetadata).IsAssignableFrom(a.AttributeType)))
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:153]: Does `OfType` not work for this?

File: src/Shared/RequestDelegateFactoryUtilities.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+                if (tryParseParameters.Length == 2 &&
+                    tryParseParameters[0].ParameterType == typeof(string) &&
+                    tryParseParameters[1].IsOut)
+                {
+                    return method;
+                }
+            }
+
+            throw new Exception("static bool System.Enum.TryParse<TEnum>(string? value, out TEnum result) does not exist!!?!?");
```
Comment (pranavkm) [inline src/Shared/RequestDelegateFactoryUtilities.cs:81]:
```suggestion
            Debug.Fail("This should never happen.");
            throw new InvalidOperationException("Unable to find Enum.TryParse");
```

File: src/Shared/RequestDelegateFactoryUtilities.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            return TryParseMethodCache.GetOrAdd(type, Finder);
+        }
+
+        private static MethodInfo GetEnumTryParseMethod()
+        {
+            var staticEnumMethods = typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static);
+
+            foreach (var method in staticEnumMethods)
+            {
+                if (!method.IsGenericMethod || method.Name != "TryParse" || method.ReturnType != typeof(bool))
```
Comment (pranavkm) [inline src/Shared/RequestDelegateFactoryUtilities.cs:66]:
```suggestion
                if (!method.IsGenericMethod || method.Name != nameof(Enum.TryParse) || method.ReturnType != typeof(bool))
```

File: src/Shared/RequestDelegateFactoryUtilities.cs [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Http
+{
+    // REVIEW: Better name?
+    internal static class RequestDelegateFactoryUtilities
```
Comment (pranavkm) [inline src/Shared/RequestDelegateFactoryUtilities.cs:11]: `TryParseMethodCache` or something like that? Naming a type utils makes it very easy for it be a dumping ground for unrelated stuff

File: src/Mvc/Mvc.ApiExplorer/test/Microsoft.AspNetCore.Mvc.ApiExplorer.Test.csproj [hunk 18]
```diff
@@ -2,6 +2,7 @@
 
   <PropertyGroup>
     <TargetFramework>$(DefaultNetCoreTargetFramework)</TargetFramework>
+    <LangVersion>Preview</LangVersion>
```
Comment (pranavkm) [inline src/Mvc/Mvc.ApiExplorer/test/Microsoft.AspNetCore.Mvc.ApiExplorer.Test.csproj:5]: ?

**Review by pranavkm — APPROVED** [review]
Happy to sign off on this. Let me know if you want me to review this further.

File: src/Mvc/Mvc.ApiExplorer/test/Microsoft.AspNetCore.Mvc.ApiExplorer.Test.csproj (hunk 18 above)
Comment (davidfowl) [reply src/Mvc/Mvc.ApiExplorer/test/Microsoft.AspNetCore.Mvc.ApiExplorer.Test.csproj:5]: Required to using C# 10

File: src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs (hunk 10 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs:20]: I was planning to discuss this as part of our normal Monday API review meeting, but I'll also send an email.

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs (hunk 12 above)
Comment (davidfowl) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:85]: That's also a valid option. We could do that and see how it shows up.

File: src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs (hunk 10 above)
Comment (Kahbazi) [reply src/Mvc/Mvc.ApiExplorer/src/DependencyInjection/EndpointMethodInfoApiExplorerServiceCollectionExtensions.cs:20]: How about `AddMinimalActionEndpointsApiExplorer`? like the type that `Map` method returns `MinimalActionEndpointConventionBuilder`.

https://github.com/dotnet/aspnetcore/blob/ce3a74883c7276e8747200ca8f06bba8b7211427/src/Http/Routing/src/Builder/MinimalActionEndpointRouteBuilderExtensions.cs#L139

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs (hunk 13 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:135]: https://github.com/dotnet/aspnetcore/issues/33644

File: src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs (hunk 14 above)
Comment (halter73) [reply src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs:153]: Yes. Technically Any is a bit better because it avoids the iterator, but it's not like this is super high perf code. If we change this we should do the same in RequestDelegateFactory. Maybe in a later cleanup.

File: src/Shared/RequestDelegateFactoryUtilities.cs (hunk 15 above)
Comment (halter73) [reply src/Shared/RequestDelegateFactoryUtilities.cs:81]: Debug.Fail sure, but why change this to InvalidOperationException?

Comment (pranavkm) [reply src/Shared/RequestDelegateFactoryUtilities.cs:81]: Idk. Force of habit?

**Comment by BrennanConroy** [comment]
API Review:
`AddMinimalActionsApiExplorer()`

That would make MinimalActions more of a public name, whereas right now it's mostly a code-name.

`AddEndpointsApiExplorer()` uses the now overloaded word "Endpoints" and it doesn't apply to all Endpoints.
